src/lib.rs

```diff
@@ -789,6 +925,107 @@ pub mod channel {
             }
         }
 
+        struct FilterMapSender<F, T, U> {
+            f: F,
+            sender: Sender<T>,
+            _p: PhantomData<U>,
+        }
+
+        impl<F, T, U> Debug for FilterMapSender<F, T, U> {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                f.debug_struct("FilterMapSender").finish_non_exhaustive()
+            }
+        }
+
+        impl<F, T, U> DynSender<U> for FilterMapSender<F, T, U>
+        where
+            F: Fn(U) -> Option<T> + Send + Sync + 'static,
+            T: Send + Sync + 'static,
+            U: Send + Sync + 'static,
+        {
+            fn send(
+                &self,
+                value: U,
+            ) -> Pin<Box<dyn Future<Output = Result<(), SendError>> + Send + '_>> {
+                Box::pin(async move {
+                    if let Some(v) = (self.f)(value) {
+                        self.sender.send(v).await
+                    } else {
+                        Ok(())
+                    }
+                })
+            }
+
+            fn try_send(
+                &self,
+                value: U,
+            ) -> Pin<Box<dyn Future<Output = Result<bool, SendError>> + Send + '_>> {
+                Box::pin(async move {
+                    if let Some(v) = (self.f)(value) {
+                        self.sender.try_send(v).await
+                    } else {
+                        Ok(true)
+                    }
+                })
+            }
+
+            fn is_rpc(&self) -> bool {
+                self.sender.is_rpc()
+            }
+
+            fn closed(&self) -> Pin<Box<dyn Future<Output = ()> + Send + Sync + '_>> {
+                match self {
+                    FilterMapSender {
+                        sender: Sender::Tokio(tx),
+                        ..
+                    } => Box::pin(tx.closed()),
+                    FilterMapSender {
+                        sender: Sender::Boxed(sink),
+                        ..
+                    } => sink.closed(),
+                }
+            }
+        }
+
+        struct FilterMapReceiver<F, T, U> {
+            f: F,
+            receiver: Receiver<T>,
+            _p: PhantomData<U>,
+        }
+
+        impl<F, T, U> Debug for FilterMapReceiver<F, T, U> {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                f.debug_struct("FilterMapReceiver").finish_non_exhaustive()
+            }
+        }
+
+        impl<F, T, U> DynReceiver<U> for FilterMapReceiver<F, T, U>
+        where
+            F: Fn(T) -> Option<U> + Send + Sync + 'static,
+            T: Send + Sync + 'static,
+            U: Send + Sync + 'static,
+        {
+            fn recv(
+                &mut self,
+            ) -> Pin<
+                Box<
```

Comment:
nit: that's a long type, maybe we add a BoxedSyncFuture type alias and use in all DynReceiver uses? But this is unrelated to this PR so can also come sometime later.